cornell-dti/course-plan
# Graduate Program Support: Onboarding and Firestore

### Summary 

This pull request adds graduate programs to Onboarding and our Firestore database, so that users can now select a graduate program (with or without undergrad). After being integrated with @vaishnavi17's requirements side code, it should be easy to get a new section on the requirements sidebar. See [Notion](https://www.notion.so/courseplan/82bf355c4b7b49f99691eb9f3afb203c?v=5da9d646f3ef424bbceb1ca0e8e00cf8&p=b248521fc9464869a61df28e5a352c16)

- [x] Added "Undergraduate" and "Graduate" sections on the first and last page of Onboarding.
- [x] Create more specific errors. The first page can be passed only if all required fields are selected, and at least one program as well. Error text now tells users exactly what they are missing. Modal can only be cancelled/clicked outside if they have a program and/or college on firebase.
- [x] Skip the transfer page if only a graduate program is selected, and delete transfer data if submitted. Update progress bar svgs to reflect this.
- [x] Uses temp graduate programs to test that everything works in Onbaording and Firebase.
- [x] Update styling on first and last Onboarding pages to reflect designs
- [x] Removed outdated Safari check from @hahnbeelee's temporary modal fix.  

![image](https://user-images.githubusercontent.com/25535093/117559323-53858c00-b052-11eb-8997-5748507dd15a.png)

![image](https://user-images.githubusercontent.com/25535093/117559342-74e67800-b052-11eb-82da-369c7818f477.png)



Remaining TODOs:

- [ ] Integrate with the requirements changes at the 3 locations I marked with TODO comments (and delete temp data)
- [ ] Make a separate PR or update this one after requirements changes are done to actually add a graduate program section to the requirements sidebar (@tcho6319 thoughts?).

### Test Plan 

1. Completely delete your firebase data, and confirm the error message that shows up mentions all missing required fields (and a degree).
2. Add an undergraduate college, add transfer credit, say you've taken the swim test, and submit the modal, and confirm everything works as before.
3. Remove the undergraduate college, select a graduate program. Confirm the transfer credit page cannot be accessed and submit. The requirements bar should be blank.
4. Open the onboarding model again, add back an undergraduate college, confirm the program you selected is still there, but that the transfer data/swim test has been deleted.
5. Submit and confirm you can have both a college and program selected at once.
6. Mess around with the required fields and confirm all error messages look good. 
7. Confirm that the x on the modal can be clicked, and the modal can be clicked outside, only when a user has a college or program on firebase.
8. Confirm that setting the firebase program to some invalid acronym (that is not TEMP, TEMP2, or TEMP3) leads to the error message @tcho6319 created about data on firebase that does not exist on this branch.
9. Check that all the new styling looks good on mobile.

### Notes 

I have a bunch of questions based on what I noticed when making these changes. Let me know what I should address!
1. Should we avoid using an image as the progress bar? This seems bad for accessibility because the text is part of the image, and not actual text that could be read by a screen reader
2. Should I update the `FirestoreCollegeOrMajorOrMinor` type name to include Program?
3. We use the style `onboarding-hidden` quite frequently in onboarding instead of using `v-if`. Should I change this like I did with the modal refactor?
4. Should the button on the first page be disabled if there is an error, so users don't think they can click it?
5. I sense an issue with the walkthrough coming up when the requirements side is added. The walkthrough depends on the user having a PE requirement, but this will not be the case for pure graduate students anymore. What do you think @einc ?

### Blockers 

- The requirements-side update coming soon
- Actual graduate program data

### Breaking Changes  



- There is now a `programs` field in `user-onboarding-data` like the `colleges` field. This should not break anything, but worth mentioning.
